### PR TITLE
Use AtomicPtr instead of AtomicUsize in Stack

### PR DESCRIPTION
**Pull request by MaloJaffre**

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! I'd imagine the motivation for a PR like this though is to reduce casts, but it looks like this increases the number of casts?

**Comment by MaloJaffre** [comment]
The real motivation is to use an AtomicPtr because an AtomicUsize doesn't appears as a casted pointer, and doesn't indicates what it references (a `Node<T>`).

**Comment by alexcrichton** [comment]
Sure yeah but I feel like this PR isn't actually leveraging the benefit of using `AtomicPtr`:

* the `_marker` field isn't removed
* Lots of `as` casts still exist

Can that be handled?

**Comment by MaloJaffre** [comment]
Sure, it's now done.

File: src/stack.rs [hunk 1]
```diff
@@ -32,18 +31,16 @@ unsafe impl<T: Sync> Sync for Drain<T> {}
...
+            head: AtomicPtr::default(),
         }
     }
 
     pub fn push(&self, data: T) {
-        let mut node = Box::new(Node { data: data, next: 0 as *mut _ });
+        let mut node = Box::new(Node { data: data, next: ptr::null_mut() });
         let mut head = self.head.load(SeqCst);
         loop {
             node.next = head as *mut _;
```
Comment (alexcrichton) [inline src/stack.rs]: I believe this `as` can be removed, right?

Comment (MaloJaffre) [inline src/stack.rs]: Oops, fixed.

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**